Bb remote debug

nkashyap opened the pull request:
- extended Aardwolf to meet blinkbox requirements
- blinkbox instruction updated
- added ErrorEvent in stringify
- optimize stringify function
- added batch file to start node js server and also moved window.console override into BB.Console.js file
- updated settings readme doc
- send response to the server even if native console is not supported
- extend console support in Aardwolf
- added ajax function to queue async ajax requests
  *

Commit:
Extended Aardwolf to meet blinkbox requirements

Commit:
fileServerBaseDir updated

Commit:
Added ErrorEvent in stringify

Commit:
Blinkbox instruction updated

Commit:
Added batch file to start node js server and also moved window.console override into BB.Console file

Commit:
updated settings readme doc

Commit:
removed unused variables and added extra conditions

Commit:
code cleanup

Commit:
added missing argument

Commit:
send response to the server even if native console is not supported

Commit:
Extend console to support code profiler

Commit:
fix to add correct function name.
e.g
var fun = function(){}
var a = {
 fun : function(){}
}
function fun(){}

all return "fun" as function name

chrismilleruk commented on js/BB.Console.js:
```diff
@@ -0,0 +1,620 @@
+window.console = (function(win){
+	var _console = win._console = win.console;
+	var userAgent = navigator.userAgent.toLowerCase();
```
Looks like you're doing browser sniffing to determine the features of various browsers. Can you find a way to do feature detection instead?

nkashyap replied:
changed the way its work now

chrismilleruk commented on js/BB.Console.js:
```diff
@@ -0,0 +1,620 @@
+	var userAgent = navigator.userAgent.toLowerCase();
+	var isFirefox = /firefox/.test(userAgent);
+	var isOpera   = /opera/.test(userAgent);
+	var isSafari  = /webkit/.test(userAgent);
```
isWebkit?

nkashyap replied:
changed the way its work now

chrismilleruk commented on js/BB.Console.js:
```diff
@@ -0,0 +1,620 @@
+	var _console = win._console = win.console;
+	var userAgent = navigator.userAgent.toLowerCase();
+	var isFirefox = /firefox/.test(userAgent);
+	var isOpera   = /opera/.test(userAgent);
```
Not used.

nkashyap replied:
changed the way its work now

chrismilleruk commented on js/BB.Console.js:
```diff
@@ -0,0 +1,620 @@
+					}
+				}                
+			}
+		}
```
These two if blocks look pretty similar, perhaps a function is required? especially if you are likely to add a third variant in the future.

nkashyap replied:
removed

chrismilleruk commented on js/BB.Console.js:
```diff
@@ -0,0 +1,620 @@
+		if (isSafari){
+			var reChromeStackItem = /^\s+at\s+(.*)((?:http|https|ftp|file):\/\/.*)$/;
+			var reChromeStackItemName = /\s*\($/;
+			var reChromeStackItemValue = /^(.+)\:(\d+\:\d+)\)?$/;
```
Difficult to review the regex here as there are no examples of the expected format present.

nkashyap replied:
changed the way its work now, stack string is now normalized based on browser mode type.

chrismilleruk commented on js/aardwolf.js:
```diff
@@ -38,7 +87,7 @@ window.Aardwolf = new (function() {
             asyncXHR.send(null);
         } catch (ex) {
-            alert('Aardwolf encountered an error while waiting for command: ' + ex.toString());
+            console.logToUI('Aardwolf encountered an error while waiting for command: ' + ex.toString());
```
Can you guarantee 100% that the console object exists at this point and that it has a function called 'logToUI'? If not, add some defensive code to your error handler.

nkashyap replied:
yes it should work now, I have added a check to see if all console methods used in this file is defined or not and if not defined its creates one.

Commit:
code cleanup, optimized and added opera stack trace support

nkashyap commented:
code cleanup, 
optimized,
additional checks added to make use custom console exist,
and opera stack trace support added (ongoing)

Commit:
Regex example string added

Regex example string added

Commit:
check addition console methods support

check addition console methods support and if not supported then assign
empty function

chrismilleruk commented on js/BB.Console.js:
```diff
@@ -0,0 +1,796 @@
+        userAgent = navigator.userAgent.toLowerCase(),
+        isFirefox = /firefox/.test(userAgent),
+        isOpera = /opera/.test(userAgent),
+        isWebkit = /webkit/.test(userAgent),
```
You are still using browser sniffing and not feature sniffing. This is considered bad practice and should be avoided.

nkashyap replied:
removed

chrismilleruk commented on js/BB.Console.js:
```diff
@@ -0,0 +1,796 @@
+        } else if (isOpera)
+        {
+            frames = updateOperaStackFrames(frames, items);
+        }
```
The above three functions are in fact the same function but with different regex being applied based on assumptions you are making as the result of your browser sniffing. What happens if they standardize and all use the same format? What happens if we meet a browser in the future which looks like none of the three above but happens to act like webkit?

Instead of browser sniffing, why don't you examine the format of the stack trace and use the most appropriate set of regexes? You could pass this set of regexes to a single function which could do the work required. One approach would be to have an array of objects, each with a test regex and  well known properties for the other regexes. You could work through these objects until you get a match and then pass the successful one to the parsing function?

c

nkashyap replied:
changed the way it works now

Commit:
Firefox Error stack string normalized

Firefox Error stack string is quite different from chrome and opera
string

Commit:
error stack trace string normalizer added

now BB.Console normalize stack trace string before posting onto the
server. it check available feature in the browser before applying
relavant normalizer

Commit:
code cleanup

Commit:
Some try & catch error handling

Commit:
Stack cleanup

getStack inner function reference removed from break point stack

Commit:
removed unnecessary carriage return

removed unnecessary carriage return from the comment block and also
removed < > brackets from anonymous function

Commit:
device breakpoint stack support added

device breakpoint stack support added

Commit:
Added try catch block and lookup formatter

Added try catch block
Added unknown error type formatter

Commit:
catch and log global errors

catch and log global errors,
check error type before stack lookup,
updated logToUI to include caller function name

Commit:
added missing semicolon

added missing semicolon

Commit:
added aardwolf stack support in console.js

added additional aardwolf stack support when error stack is not
available in console.js

Commit:
Update code to support samsung TV

chrismilleruk commented on README.md:
```diff
@@ -1,3 +1,14 @@
+Blinkbox Setting Aardwolf locally
```
This is one for the wiki, not public source. Can you remove please?

chrismilleruk commented on config/config.local.js:
```diff
@@ -0,0 +1,9 @@
+var config = require('../config/config.defaults.js'); 
+
+/* Hostname or IP of the local machine */
+config.serverHost = 'tv-thedictator.bbdev1.com';
```
No internal network details in public source pls.

chrismilleruk commented on js/BB.Console.js:
```diff
@@ -0,0 +1,818 @@
+		return 'other';
+	};
+
+	var formatter = {
```
This still feels a bit repetitive. I'm sure there's a much nicer way to do this and I'm not sure it covers the various devices well enough. Perhaps we can have a catch up about this in the morning...

Commit:
blinkbox references removed

Commit:
Aardwolf check added
